### PR TITLE
Ensemble calibrate

### LINKED ISSUE #435: Add `calibrate ensemble` interface
See description

### PR DESCRIPTION
**Pull request by SamWitty**
closes #435

### CODE REVIEW

File: pyciemss/interfaces.py [hunk 1]
```diff
@@ -148,6 +149,105 @@ def wrapped_model():
...
+    ],
+    data_path: str,
+    *,
+    dirichlet_alpha: Optional[torch.Tensor] = None,
+    data_mapping: Dict[str, str] = {},
+    noise_model: str = "normal",
+    noise_model_kwargs: Dict[str, Any] = {"scale": 0.1},
+    solver_method: str = "dopri5",
+    solver_options: Dict[str, Any] = {},
+    start_time: float = 0.0,
```
Comment (djinnome) [inline pyciemss/interfaces.py:166]: Should we include these parameters?

static_state_interventions: Dict[float, Dict[str, Intervention]]
            - A dictionary of static interventions to apply to the model.
            - Each key is the time at which the intervention is applied.
            - Each value is a dictionary of the form {state_variable_name: intervention_assignment}.
            - Note that the `intervention_assignment` can be any type supported by
              :func:`~chirho.interventional.ops.intervene`, including functions.
        static_parameter_interventions: Dict[float, Dict[str, Intervention]]
            - A dictionary of static interventions to apply to the model.
            - Each key is the time at which the intervention is applied.
            - Each value is a dictionary of the form {parameter_name: intervention_assignment}.
            - Note that the `intervention_assignment` can be any type supported by
              :func:`~chirho.interventional.ops.intervene`, including functions.
        dynamic_state_interventions: Dict[
                                        Callable[[torch.Tensor, Dict[str, torch.Tensor]], torch.Tensor],
                                        Dict[str, Intervention]
                                        ]
            - A dictionary of dynamic interventions to apply to the model.
            - Each key is a function that takes in the current state of the model and returns a tensor.
              When this function crosses 0, the dynamic intervention is applied.
            - Each value is a dictionary of the form {state_variable_name: intervention_assignment}.
            - Note that the `intervention_assignment` can be any type supported by
              :func:`~chirho.interventional.ops.intervene`, including functions.

File: pyciemss/interfaces.py [hunk 2]
```diff
@@ -148,6 +149,105 @@ def wrapped_model():
...
+    num_iterations: int = 1000,
+    lr: float = 0.03,
+    verbose: bool = False,
+    num_particles: int = 1,
+    deterministic_learnable_parameters: List[str] = [],
+    progress_hook: Callable = lambda i, loss: None,
+) -> Dict[str, Any]:
```
Comment (djinnome) [inline pyciemss/interfaces.py:173]: Would you be willing to add a docstring, something like this:

"""
    Load a collection of models from files, compile them into an ensemble probabilistic program, and Infer parameters for the ensemble system conditional on data.
    This uses variational inference with a mean-field variational family to infer the parameters of the model.
Args:
    model_paths_or_jsons: List[Union[str, Dict]]
        - A list of paths to AMR model files or JSONs containing models in AMR form.
    data_path: str
            - A path to the data file.
    dirichlet_alpha: Optional[torch.Tensor]
        - A tensor of shape (num_models,) containing the Dirichlet alpha values for the ensemble.
            - A higher proportion of alpha values will result in higher weights for the corresponding models.
            - A larger total alpha values will result in more certain priors.
            - e.g. torch.tensor([1, 1, 1]) will result in a uniform prior over vectors of length 3 that sum to 1.
            - e.g. torch.tensor([1, 2, 3]) will result in a prior that is biased towards the third model.
        - If not provided, we will use a uniform Dirichlet prior.
     data_mapping: Dict[str, str]
            - A mapping from column names in the data file to state variable names in the model.
                - keys: str name of column in dataset
                - values: str name of state/observable in model
            - If not provided, we will assume that the column names in the data file match the state variable names.
      noise_model: str
            - The noise model to use for the data.
            - Currently we only support the normal distribution.
      noise_model_kwargs: Dict[str, Any]
            - Keyword arguments to pass to the noise model.
            - Currently we only support the `scale` keyword argument for the normal distribution.
      solver_method: str
            - The method to use for solving the ODE. See torchdiffeq's `odeint` method for more details.
            - If performance is incredibly slow, we suggest using `euler` to debug.
              If using `euler` results in faster simulation, the issue is likely that the model is stiff.
      solver_options: Dict[str, Any]
            - Options to pass to the solver. See torchdiffeq' `odeint` method for more details.
      start_time: float
            - The start time of the model. This is used to align the `start_state` from the
              AMR model with the simulation timepoints.
            - By default we set the `start_time` to be 0.
            - - num_iterations: int
            - The number of iterations to run the inference algorithm for.
       lr: float
            - The learning rate to use for the inference algorithm.
      verbose: bool
            - Whether to print out the loss at each iteration.
       num_particles: int
            - The number of particles to use for the inference algorithm.
       deterministic_learnable_parameters: List[str]
            - A list of parameter names that should be learned deterministically.
            - By default, all parameters are learned probabilistically.
      progress_hook: Callable[[int, float], None]
            - A function that takes in the current iteration and the current loss.
            - This is called at the beginning of each iteration.
            - By default, this is a no-op.
            - This can be used to implement custom progress bars.

**Review by djinnome — COMMENTED** [review]
Looks good except for the fact that ensemble_calibrate has no docstring

Also, do we want to include `static_state_interventions, static_parameter_interventions, dynamic_state_interventions`, and `dynamic_parameter_interventions.`

**Comment by SamWitty** [comment]
@djinnome . This is a WIP draft PR, and is not yet ready for review. Perhaps you meant to review #523 ?

File: pyciemss/interfaces.py (hunk 1 above)
Comment (SamWitty) [reply pyciemss/interfaces.py:166]: No, we currently do not support interventions with ensemble models. Doing so would require that the solution mappings are bijective, which is hard to check in general. The challenge is that when a user describes an intervention on the ensemble, we would need to propagate the meaning of that intervention into each of the constituent models.
